### ISSUE TITLE
clean exits

### ISSUE THREAD

**Issue by michaelcushing** [pyspextool/pyspextool#129]
So this came up in our last call.  When things break, it would be nice if it would do so "nicely" so that things can continue.  So for a concrete example, let's imagine something breaks at the combine_spectra routines.  It would be nice for the software to return something that we can test and say "ok, that broke, skip the rest and keep going".

In IDL, I accomplished this with a keyword CANCEL.

If anything went wrong in a function, I would set CANCEL to 1, and then return from the function.  The next line after any function call was always,

doingsomething(data, CANCEL=cancel)
if cancel then return

I could do this all the way out of nested calls until I get to a nice safe place.

When I started coding pyspextool, I didn't see an obvious way to do this given that one cannot return values through a keyword.  I considered returning a "cancel" variable for every function but that felt non-pythonic and so didn't bother.  

So once the try/except fails, how do we tell the script/function that called the routine with the try/except that things failed and that one should not try and continue?

**Comment by kelle** [comment]
Can you be even more concrete? What type of thing might happen during the combine_spectra step?

Also, when you say "continue", do you mean like, keep going to the next set of spectra? Or do you mean you just want to fail at a higher level function?  I see the need for both uses, just want to make sure we're on the same page about the one we're talking about now.

**Comment by michaelcushing** [comment]
For the moment I am thinking about Adam's batch runs.  

So if we get some bad data or something, it would be nice for whatever routine breaks (making flats, or extraction, or telluric correction) to return to his batch script with something that allows his batch script to say "ok, this set broke, let's skip what remains of this data set and move on to the next one"

**Comment by kelle** [comment]
I have implemented this in SIMPLE ingests scripts with a CONTINUE keyword. If false, it stops at an error. If true, it communicates about the error but then keeps on trucking. 

'''Python 
Try:
   do_file(file)
Except(OSError) as e:
   If continue:
      Logging.warning("something didn't work. Skipping file")
   Continue 
   Else:
      Logging.error(f"something didn't work. Skipping {file}")
   Return(e)
'''

**Comment by kelle** [comment]
ok, I called it `raise_error`
```
raise_error: bool, optional
        True (default): Raise an error if a source cannot be ingested
        False: Log a warning but skip sources which cannot be ingested
```
Examples:
```python
simbad_result_table = Simbad.query_object(source)
if simbad_result_table is None:
   n_skipped += 1
   msg = f"{i}: Skipping: {source}. Coordinates are needed and could not be retrieved from SIMBAD. \n"
   logger.warning(msg)
   if raise_error:
           raise SimpleError(msg)
   else:
         continue
```
```python
try:
   ingest_names(db, source, source)
   n_names += 1
except sqlalchemy.exc.IntegrityError:
    msg = f"{i}: Could not add {names_data} to database"
    logger.warning(msg)
    if raise_error:
        raise SimpleError(msg)
    else:
        continue
```

**Comment by michaelcushing** [comment]
I see.  So Adam could in principle add try/excepts for each function call in the loop, and if it breaks, he can just skip to the next interaction of the loop and log it?

**Comment by kelle** [comment]
Yep.

**Comment by michaelcushing** [comment]
@aburgasser

**Comment by michaelcushing** [comment]
@Adam.  Let me poke at this as I think I can accomplish what we need inside the core code.

**Comment by michaelcushing** [comment]
So in working on #107, I found a perfect example to try and test this.  One of the pairs is "bad" and so there aren't two apertures, and when it goes to generate an aperture mask (which labels each pixel in the slit with a number describing its purpose - aperture, background, etc.) with two apertures it breaks.  So my goal was to exit define_aperture_parameters nicely so as to continue to the do_all_steps program.

I added a try/except around define_aperture_parameters in do_all_steps:

       try:
            
            define_aperture_parameters(extract.parameters['apradii'],
                                   psf_radius=extract.parameters['psfradius'],
                                   bg_radius=extract.parameters['bgradius'],
                                   bg_width=extract.parameters['bgwidth'],
                                   bg_regions=extract.parameters['bgregions'],
                                   bg_fit_degree=extract.parameters['bgdeg'],
                                   qa_show=extract.parameters['qaplot'],
                                   qa_showsize=extract.parameters['qaplotsize'],
                                   qa_write=extract.parameters['qafile'])

        except(ValueError) as e:
            
            continue

and that worked well (will fill in more in the except(ValueError) section later: logging, messages, etc.)

In order to relay this information to the do_all_steps caller, I was planning on return a list of the successful files.  so for example, if the user passes '69-72', and 71,72 are the ones that fail, do_all_steps would return a list of [69,70].

Then the batch code would have to adjust its call to say combine_spectra to skip 71 and 72.  eventually I can wrap each call in do_all_steps in something similar to try and catch all the problems.

Does this sound reasonable?  And if so, @aburgasser what is the best format for the returned files for you?  Also, one of course has to run everything a first time before calling do_all_steps and so we have to include those files as well, but that should be straightforward.
